### PR TITLE
rgw/http/notifications: support content type in HTTP POST messages

### PR DESCRIPTION
**Pull request by yuvalif**
Fixes: https://tracker.ceph.com/issues/51530

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_http_client.cc [hunk 1]
```diff
@@ -600,6 +600,10 @@ int RGWHTTPClient::init_request(rgw_http_req_data *_req_data)
       // TODO: set to size smaller than 1MB should prevent the "Expect" field
       // from being sent. So explicit removal is not needed
       h = curl_slist_append(h, "Expect:");
+      if (!content_type.empty()) {
+        // set content type in case of POST
+        h = curl_slist_append(h, content_type.c_str());
+      }
```
Comment (cbodley) [inline src/rgw/rgw_http_client.cc:606]: why does this need to be specific to `method == "POST"`?

can we just pass in the `Content-Type` as a normal header? keystone auth is doing this with:
```c++
token_req.append_header("Content-Type", "application/json");
```

Comment (yuvalif) [reply src/rgw/rgw_http_client.cc:606]: ok. will add as a regular header

Comment (yuvalif) [reply src/rgw/rgw_http_client.cc]: added regardless of the method is "POST".
not sure what is the difference between the two ways of adding to the header?

File: src/rgw/rgw_pubsub_push.cc [hunk 2]
```diff
@@ -141,6 +141,7 @@ class RGWPubSubHTTPEndpoint : public RGWPubSubEndpoint {
     const auto post_data = json_format_pubsub_event(event);
     request.set_post_data(post_data);
     request.set_send_length(post_data.length());
+    request.set_content_type("application/json");
```
Comment (cbodley) [inline src/rgw/rgw_pubsub_push.cc:144]: can't this just be:
```cpp
request.append_header("Content-Type", "application/json");
```
i'd rather not to add a separate interface specifically for content-type

**Comment by yuvalif** [comment]
teuthology results see here: https://github.com/ceph/ceph/pull/42355#issuecomment-882407021

**Comment by yuvalif** [comment]
> teuthology results see here: [#42355 (comment)](https://github.com/ceph/ceph/pull/42355#issuecomment-882407021)

failures are known issues.
bucket notification tests are now passing: http://pulpito.front.sepia.ceph.com/yuvalif-2021-07-19_14:52:02-rgw:notifications-wip-yuval-testing1-distro-basic-smithi/

**Review by cbodley — APPROVED** [review]

**Pull request merged**